flink-runtime/src/main/java/org/apache/flink/runtime/registration/RegistrationResponse.java

```diff
@@ -42,30 +44,42 @@ public String toString() {
 
     // ----------------------------------------------------------------------------
 
-    /** A rejected (declined) registration. */
-    public static final class Decline extends RegistrationResponse {
+    /** A registration failure. */
+    public static final class Failure extends RegistrationResponse {
         private static final long serialVersionUID = 1L;
 
-        /** The rejection reason. */
-        private final String reason;
+        /** The failure reason. */
+        private final SerializedThrowable reason;
 
         /**
-         * Creates a new rejection message.
+         * Creates a new failure message.
          *
-         * @param reason The reason for the rejection.
+         * @param reason The reason for the failure.
          */
-        public Decline(String reason) {
-            this.reason = reason != null ? reason : "(unknown)";
+        public Failure(Throwable reason) {
+            this.reason = new SerializedThrowable(reason);
         }
 
-        /** Gets the reason for the rejection. */
-        public String getReason() {
+        /** Gets the reason for the failure. */
+        public SerializedThrowable getReason() {
             return reason;
         }
 
         @Override
         public String toString() {
-            return "Registration Declined (" + reason + ')';
+            return "Registration Failure (" + reason + ')';
+        }
+    }
+
+    // ----------------------------------------------------------------------------
+
+    /** A rejected (declined) registration. */
```

Comment:
We should clarify in what circumstances a Rejection or Failure should be returned.

Reply:
Good point. I will add details to their JavaDocs.